File: src/MMORPG.GodotClient/Scripts/Visuals/LpcPaperdollEngine.cs
using Godot;
using System;
using System.IO;
using System.Collections.Generic;
using MMORPG.Shared.Enums;
using MMORPG.Shared.Registry;

public static class LpcPaperdollEngine
{
    private const int FrameSize = 64;

    private static readonly Dictionary<string, int> WalkRowY = new()
    {
        ["north"] = 512,
        ["west"] = 576,
        ["south"] = 640,
        ["east"] = 704
    };

    private static readonly Dictionary<string, string> DiagMapping = new()
    {
        ["south"] = "south",
        ["south-east"] = "east",
        ["east"] = "east",
        ["north-east"] = "north",
        ["north"] = "north",
        ["north-west"] = "west",
        ["west"] = "west",
        ["south-west"] = "south"
    };

    public static void ExtractLpcEquipmentSheets()
    {
        string projectRoot = ProjectSettings.GlobalizePath("res://");
        string repoBase = Path.Combine(projectRoot, "../../tools/lpc_generator/spritesheets");
        string outputBase = Path.Combine(projectRoot, "Assets/Textures/Paperdoll/LPC");

        if (!Directory.Exists(repoBase))
        {
            GD.PrintErr($"[LPC Engine] Local LPC repository not found at '{repoBase}'. Skipping extraction.");
            return;
        }

        var itemsToExtract = new Dictionary<string, string>
        {
            ["Head/IronHelm"] = FindFirstFile(repoBase, "head", "helmets", "*.png"),
            ["Armor/IronPlateChest"] = FindFirstFile(repoBase, "torso", "armors", "*.png"),
            ["Legs/IronLeggings"] = FindFirstFile(repoBase, "legs", "pants", "*.png"),
            ["Boots/IronBoots"] = FindFirstFile(repoBase, "feet", "boots", "*.png"),
            ["Weapons/IronSword"] = FindFirstFile(repoBase, "weapon", "sword", "*.png")
        };

        foreach (var (itemKey, sheetPath) in itemsToExtract)
        {
            if (string.IsNullOrEmpty(sheetPath) || !File.Exists(sheetPath))
            {
                GD.PrintErr($"[LPC Engine] Could not find sheet for '{itemKey}'.");
                continue;
            }

            string outDir = Path.Combine(outputBase, itemKey);
            Directory.CreateDirectory(outDir);
            ExtractSheet(sheetPath, outDir, itemKey);
        }
    }

    private static string FindFirstFile(string baseDir, string sub1, string sub2, string pattern)
    {
        string targetDir = Path.Combine(baseDir, sub1, sub2);
        if (!Directory.Exists(targetDir)) targetDir = Path.Combine(baseDir, sub1);
        if (!Directory.Exists(targetDir)) return string.Empty;

        string[] files = Directory.GetFiles(targetDir, pattern, SearchOption.AllDirectories);
        return files.Length > 0 ? files[0] : string.Empty;
    }

    private static void ExtractSheet(string sheetPath, string outDir, string itemKey)
    {
        Image fullSheet = Image.LoadFromFile(sheetPath);
        if (fullSheet == null) return;

        foreach (var (dirName, startY) in WalkRowY)
        {
            int y = startY;
            if (y + FrameSize > fullSheet.GetHeight()) y = 0;

            Image frameImg = fullSheet.GetRegion(new Rect2I(0, y, FrameSize, FrameSize));
            string framePath = Path.Combine(outDir, $"{dirName}.png");
            frameImg.SavePng(framePath);
            GD.Print($"[LPC Engine] Successfully extracted '{itemKey}' ({dirName}) -> {framePath}");
        }

        // Copy diagonal aliases
        foreach (var (diagDir, targetDir) in DiagMapping)
        {
            if (WalkRowY.ContainsKey(diagDir)) continue;
            string srcPath = Path.Combine(outDir, $"{targetDir}.png");
            string destPath = Path.Combine(outDir, $"{diagDir}.png");
            if (File.Exists(srcPath) && !File.Exists(destPath))
            {
                File.Copy(srcPath, destPath, true);
            }
        }
    }
}

File: scripts/lpc_paperdoll_engine.py
import os
import glob
from PIL import Image

# LPC Standard 64x64 Grid Sheet Format
# Row 0 (Y: 512px): North (9 frames)
# Row 1 (Y: 576px): West (9 frames)
# Row 2 (Y: 640px): South (9 frames)
# Row 3 (Y: 704px): East (9 frames)

FRAME_SIZE = 64
WALK_ROW_Y = {
    "north": 512,
    "west": 576,
    "south": 640,
    "east": 704
}

# Diagonal mappings to cardinal nearest
DIR_MAPPING = {
    "south": "south",
    "south-east": "east",
    "east": "east",
    "north-east": "north",
    "north": "north",
    "north-west": "west",
    "west": "west",
    "south-west": "south"
}

def extract_lpc_walk_frames(sheet_path, output_dir, item_name):
    if not os.path.exists(sheet_path):
        print(f"Sheet path not found: {sheet_path}")
        return False

    img = Image.open(sheet_path).convert("RGBA")
    os.makedirs(output_dir, exist_ok=True)

    # Extract 4 cardinal walk directions (north, west, south, east)
    for dir_name, start_y in WALK_ROW_Y.items():
        if start_y + FRAME_SIZE > img.height:
            # Fallback if image has different Y offsets
            start_y = 0

        # Frame 0 is resting idle, Frame 1..8 are walk steps
        crop_box = (0, start_y, FRAME_SIZE, start_y + FRAME_SIZE)
        frame_img = img.crop(crop_box)
        frame_path = os.path.join(output_dir, f"{dir_name}.png")
        frame_img.save(frame_path, "PNG")
        print(f"Extracted LPC layer '{item_name}' ({dir_name}): {frame_path}")

    # Create diagonal direction aliases (south-east, north-east, north-west, south-west)
    for diag_dir, cardinal_target in DIR_MAPPING.items():
        if diag_dir in WALK_ROW_Y:
            continue
        src_path = os.path.join(output_dir, f"{cardinal_target}.png")
        dest_path = os.path.join(output_dir, f"{diag_dir}.png")
        if os.path.exists(src_path) and not os.path.exists(dest_path):
            img_cardinal = Image.open(src_path)
            img_cardinal.save(dest_path, "PNG")

    return True

def main():
    repo_base = "tools/lpc_generator/spritesheets"
    output_base = "src/MMORPG.GodotClient/Assets/Textures/Paperdoll/LPC"

    items_to_process = {
        "Head/IronHelm": glob.glob(os.path.join(repo_base, "head/helmets/**/walk/*.png"), recursive=True) + glob.glob(os.path.join(repo_base, "head/helmets/**/*.png"), recursive=True),
        "Armor/IronPlateChest": glob.glob(os.path.join(repo_base, "torso/armors/**/walk/*.png"), recursive=True) + glob.glob(os.path.join(repo_base, "torso/armors/**/*.png"), recursive=True) + glob.glob(os.path.join(repo_base, "torso/**/plate*.png"), recursive=True),
        "Legs/IronLeggings": glob.glob(os.path.join(repo_base, "legs/pants/**/walk/*.png"), recursive=True) + glob.glob(os.path.join(repo_base, "legs/**/*.png"), recursive=True),
        "Boots/IronBoots": glob.glob(os.path.join(repo_base, "feet/boots/**/walk/*.png"), recursive=True) + glob.glob(os.path.join(repo_base, "feet/**/*.png"), recursive=True),
        "Weapons/IronSword": glob.glob(os.path.join(repo_base, "weapon/sword/**/walk/*.png"), recursive=True) + glob.glob(os.path.join(repo_base, "weapon/sword/**/*.png"), recursive=True)
    }

    for item_key, matches in items_to_process.items():
        if matches:
            chosen_sheet = matches[0]
            out_dir = os.path.join(output_base, item_key)
            extract_lpc_walk_frames(chosen_sheet, out_dir, item_key)
        else:
            print(f"No LPC matching sheet found for {item_key}")

if __name__ == "__main__":
    main()
